File: android/app/src/main/java/com/attentionmirror/domain/EstimateEngine.kt
package com.attentionmirror.domain

import java.math.BigDecimal
import java.math.RoundingMode

/** Estimate for a single platform over a usage window. */
data class PlatformEstimate(
    val platform: String,
    val packageName: String,
    val minutes: Double,
    val estimatedAdsSeen: Int,
    val valueLowInr: Double,
    val valueHighInr: Double,
)

/** Aggregated, honest estimate across all monetized platforms. */
data class AttentionReceipt(
    val totalMinutes: Double,
    val estimatedAdsSeen: Int,
    val estimatedValueLowInr: Int,
    val estimatedValueHighInr: Int,
    val userReceivedInr: Int,
    val perPlatform: List<PlatformEstimate>,
)

/**
 * Canonical attention-value math. Mirrors `backend/app/estimate.py`. Everything
 * here is a transparent estimate — never an exact claim. See
 * `docs/ESTIMATE_SPEC.md`.
 */
object EstimateEngine {

    private fun roundHalfUp(value: Double): Int =
        BigDecimal.valueOf(value).setScale(0, RoundingMode.HALF_UP).toInt()

    /**
     * @param personalAdsPerMinute when non-null, overrides the platform default
     *   (used for user-assisted calibration; see [Calibration]).
     */
    fun estimatePlatform(
        config: PlatformConfig,
        durationSeconds: Long,
        personalAdsPerMinute: Double? = null,
    ): PlatformEstimate {
        val seconds = durationSeconds.coerceAtLeast(0L)
        val minutes = seconds / 60.0

        if (!config.monetized) {
            return PlatformEstimate(
                platform = config.platform,
                packageName = config.packageName,
                minutes = minutes,
                estimatedAdsSeen = 0,
                valueLowInr = 0.0,
                valueHighInr = 0.0,
            )
        }

        val rate = personalAdsPerMinute ?: config.adsPerMinute
        val ads = roundHalfUp(minutes * rate)
        return PlatformEstimate(
            platform = config.platform,
            packageName = config.packageName,
            minutes = minutes,
            estimatedAdsSeen = ads,
            valueLowInr = ads * config.lowCpmInr / 1000.0,
            valueHighInr = ads * config.highCpmInr / 1000.0,
        )
    }

    /**
     * Combine per-package usage (seconds) into a daily/weekly receipt. Packages
     * without a known config are ignored entirely.
     */
    fun buildReceipt(
        configsByPackage: Map<String, PlatformConfig>,
        usageSecondsByPackage: Map<String, Long>,
        personalRatesByPackage: Map<String, Double> = emptyMap(),
    ): AttentionReceipt {
        var totalMinutes = 0.0
        var totalAds = 0
        var totalLow = 0.0
        var totalHigh = 0.0
        val perPlatform = mutableListOf<PlatformEstimate>()

        for ((pkg, seconds) in usageSecondsByPackage) {
            val config = configsByPackage[pkg] ?: continue
            val est = estimatePlatform(config, seconds, personalRatesByPackage[pkg])
            perPlatform += est
            totalMinutes += est.minutes
            totalAds += est.estimatedAdsSeen
            totalLow += est.valueLowInr
            totalHigh += est.valueHighInr
        }

        perPlatform.sortByDescending { it.minutes }

        return AttentionReceipt(
            totalMinutes = totalMinutes,
            estimatedAdsSeen = totalAds,
            estimatedValueLowInr = roundHalfUp(totalLow),
            estimatedValueHighInr = roundHalfUp(totalHigh),
            userReceivedInr = 0,
            perPlatform = perPlatform,
        )
    }
}

File: ios/Shared/EstimateEngine.swift
import Foundation

// Portable attention-value math for the (future) iOS app. Mirrors
// android `domain/EstimateEngine.kt` and `backend/app/estimate.py` exactly —
// keep all three in sync with docs/ESTIMATE_SPEC.md. Everything here is a
// transparent estimate, never an exact claim.

public struct PlatformConfig {
    public let platform: String
    public let bundleId: String
    public let adsPerMinute: Double
    public let lowCpmInr: Double
    public let highCpmInr: Double
    public let monetized: Bool

    public init(platform: String, bundleId: String, adsPerMinute: Double,
                lowCpmInr: Double, highCpmInr: Double, monetized: Bool = true) {
        self.platform = platform
        self.bundleId = bundleId
        self.adsPerMinute = adsPerMinute
        self.lowCpmInr = lowCpmInr
        self.highCpmInr = highCpmInr
        self.monetized = monetized
    }
}

public struct PlatformEstimate {
    public let platform: String
    public let bundleId: String
    public let minutes: Double
    public let estimatedAdsSeen: Int
    public let valueLowInr: Double
    public let valueHighInr: Double
}

public struct AttentionReceipt {
    public let totalMinutes: Double
    public let estimatedAdsSeen: Int
    public let estimatedValueLowInr: Int
    public let estimatedValueHighInr: Int
    public let userReceivedInr: Int
    public let perPlatform: [PlatformEstimate]
}

public enum EstimateEngine {

    private static func roundHalfUp(_ value: Double) -> Int {
        // Values are non-negative, so away-from-zero == half-up.
        Int(value.rounded(.toNearestOrAwayFromZero))
    }

    public static func estimatePlatform(
        config: PlatformConfig,
        durationSeconds: Int,
        personalAdsPerMinute: Double? = nil
    ) -> PlatformEstimate {
        let seconds = max(0, durationSeconds)
        let minutes = Double(seconds) / 60.0

        if !config.monetized {
            return PlatformEstimate(platform: config.platform, bundleId: config.bundleId,
                                    minutes: minutes, estimatedAdsSeen: 0,
                                    valueLowInr: 0, valueHighInr: 0)
        }

        let rate = personalAdsPerMinute ?? config.adsPerMinute
        let ads = roundHalfUp(minutes * rate)
        return PlatformEstimate(
            platform: config.platform,
            bundleId: config.bundleId,
            minutes: minutes,
            estimatedAdsSeen: ads,
            valueLowInr: Double(ads) * config.lowCpmInr / 1000.0,
            valueHighInr: Double(ads) * config.highCpmInr / 1000.0
        )
    }

    public static func buildReceipt(
        configsByBundle: [String: PlatformConfig],
        usageSecondsByBundle: [String: Int],
        personalRatesByBundle: [String: Double] = [:]
    ) -> AttentionReceipt {
        var totalMinutes = 0.0, totalLow = 0.0, totalHigh = 0.0
        var totalAds = 0
        var perPlatform: [PlatformEstimate] = []

        for (bundle, seconds) in usageSecondsByBundle {
            guard let config = configsByBundle[bundle] else { continue }
            let est = estimatePlatform(config: config, durationSeconds: seconds,
                                       personalAdsPerMinute: personalRatesByBundle[bundle])
            perPlatform.append(est)
            totalMinutes += est.minutes
            totalAds += est.estimatedAdsSeen
            totalLow += est.valueLowInr
            totalHigh += est.valueHighInr
        }

        perPlatform.sort { $0.minutes > $1.minutes }

        return AttentionReceipt(
            totalMinutes: totalMinutes,
            estimatedAdsSeen: totalAds,
            estimatedValueLowInr: roundHalfUp(totalLow),
            estimatedValueHighInr: roundHalfUp(totalHigh),
            userReceivedInr: 0,
            perPlatform: perPlatform
        )
    }
}
